cpu/sam0_common/periph/freqm.c

```diff
@@ -0,0 +1,224 @@
+}
+
+typedef struct {
+    freqm_callback_t callback;
+    void *context;
+} freqm_object_t;
+
+static volatile freqm_object_t freqm_obj;
+
+static mutex_t sync_lock = MUTEX_INIT_LOCKED;
+
+static void _freqm_enable(uint8_t refnum)
+{
+    FREQM->CFGA.reg = (uint16_t)(FREQM_CFGA_REFNUM(refnum));
+
+    /* Enable DONE Interrupt */
+    FREQM->INTENSET.reg = (uint8_t)FREQM_INTENSET_DONE_Msk;
+
+    /* Enable FREQM */
+    FREQM->CTRLA.reg = (uint8_t)FREQM_CTRLA_ENABLE_Msk;
+
+    /* Wait for Sync */
+    while ((FREQM->SYNCBUSY.reg) != 0U) {}
+}
+
+static bool _freqm_is_busy(void)
+{
+    bool freqm_is_busy = false;
+
+    if ((FREQM->STATUS.reg & FREQM_STATUS_BUSY_Msk) == FREQM_STATUS_BUSY_Msk) {
+        freqm_is_busy = true;
+    }
+
+    return freqm_is_busy;
+}
+
+static void _freqm_nvic_init(void)
+{
+    /* Enable interrupt */
+    NVIC_EnableIRQ(FREQM_IRQn);
+}
+
+bool freqm_start(int blocking)
+{
+    bool status = false;
+
+    /* Check if measurement is already in progress */
+    if (!_freqm_is_busy()) {
+        /* Clear the Done Interrupt flag */
+        FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+        /* Start measurement */
+        FREQM->CTRLB.reg = (uint8_t)FREQM_CTRLB_START_Msk;
+
+        status = true;
+    }
+
+    if (blocking) {
+        /* Block until measurement is done */
+        mutex_lock(&sync_lock);
+    }
+
+    return status;
+}
+
+int freqm_frequency_get(const struct freqm_config *config, uint32_t *result)
+{
+    uint64_t result_tmp = 0U;
+    int overflow_condition = 0;
+
+    /* Calculate measured frequency */
+    result_tmp = FREQM->VALUE.reg * (uint64_t)(sam0_gclk_freq(config->ref_clk_src));
+    result_tmp = result_tmp / config->ref_clk_cnt;
+
+    /* Return measured frequency */
+    *result = (uint32_t)result_tmp;
+
+    /* Read overflow status */
+    overflow_condition = ((int)FREQM->STATUS.reg & FREQM_STATUS_OVF_Msk);
+
+    /* Clear overflow status */
+    FREQM->STATUS.reg = (uint8_t)FREQM_STATUS_OVF_Msk;
+
+    return overflow_condition;
+}
+
+void freqm_callback_register(freqm_callback_t freqm_callback, void *context)
+{
+    freqm_obj.callback = freqm_callback;
+
+    freqm_obj.context = context;
+}
+
+void irq_freqm(void)
+{
+    void *context_var;
+
+    FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+    if (freqm_obj.callback != NULL) {
+        if (_freqm_is_busy() == false) {
+            context_var = freqm_obj.context;
+            freqm_obj.callback(context_var);
+        }
+    }
+
+    mutex_unlock(&sync_lock);
+}
+
+int freqm_init(struct freqm_config const *config)
+{
+    uint8_t gclk_io_id;
+
+    /* Sanity check configuration */
+    if (config->ref_clk_src > GCLK_GEN_NUM_MSB) {
+        return -1;
+    }
+
+    if (_freqm_gpio_init(config->msr_clk_src, &gclk_io_id) != 0) {
+        return -1;
+    }
+
+    _freqm_clock_init(gclk_io_id, config->ref_clk_src);
+    _freqm_enable(config->ref_clk_cnt);
```

Comment:
Would it make sense to only do this in `freqm_start()` to save power?

Reply:
I couldn't find any info on the power consumption after enabling the peripheral. But I agree it makes more sense putting this into `freqm_start()` (or `freqm_frequency_get()`, if the functions get joined) and disabling it after the measurement.

Reply:
You can remove the current measure jumper on the `same54-xpro` and connect a multi-meter instead. You will see that usually when enabling a peripheral, current consumption increases by a few mA.